src/ipc/ipc-common.c

```diff
@@ -212,9 +228,30 @@ void ipc_msg_send(struct ipc_msg *msg, void *data, bool high_priority)
 			list_item_append(&msg->list, &ipc->msg_list);
 	}
 
+	schedule_ipc_worker();
+
+out:
+	k_spin_unlock(&ipc->lock, key);
+}
+
+#ifdef __ZEPHYR__
+static void ipc_work_handler(struct k_work *work)
+{
+	struct ipc *ipc = ipc_get();
+	k_spinlock_key_t key;
+
+	ipc_send_queued_msg();
+
+	key = k_spin_lock(&ipc->lock);
+	if (list_is_empty(&ipc->msg_list) || ipc->pm_prepare_D3)
+		goto out;
+
+	schedule_ipc_worker();
```

Comment:
sorry for a late comment, but how about
```
if (!list_is_empty(&ipc->msg_list) || ipc->pm_prepare_D3)
	schedule_ipc_worker();
```
and removing the `goto` and the label?

Reply:
@lyakh I actually put this in just for you :D not to have a double-negative check. The full check needs to be 

```
if (!list_is_empty(&ipc->msg_list) && !ipc->pm_prepare_D3)
	schedule_ipc_worker();
```

I actually prefer that, so I'll change. 

```
if (!(list_is_empty(&ipc->msg_list) || ipc->pm_prepare_D3))
	schedule_ipc_worker();
```

... I find that harder to read myself.

Reply:
@lyakh, I think this would be the correct check:
```c
if (!list_is_empty(&ipc->msg_list) && !ipc->pm_prepare_D3)
	schedule_ipc_worker();
```

Reply:
V3 update:
 - goto out, double negative check in
 - no changes otherwise

Reply:
@kv2019i @ujfalusi oops, missed that the `if` had two conditions in it, sorry, you're certainly right with the correction

Reply:
> @lyakh I actually put this in just for you :D

:-D sure, where removing a negation *simplifies* the code I'm all for it, in cases like
```
if (!a)
	x();
else
	y();
```
and similar. But without an `else` I usually prefer a negation over a `goto`. Of course, there are exceptions: for error handling, when the label is already there etc.

Reply:
My rule of thumb for deciding between `!(A || B)` versus `!A && !B` is: how closely related are A and B? If they're mostly independent of each then `!A && !B` is more readable, otherwise it's `!(A || B)`. I think this approach makes the code closer to plain English / more readable.

My 2 cents; as long as there is no double negation then everything is fine :-)